C7Engine/EntryPoints/AnimationTracker.cs

```diff
@@ -0,0 +1,99 @@
+// pulls in a lot of additional baggage. Specifically it means the engine must know about real world time, unit INI files, and the player's interface
+// settings.
+
+// The callbacks are hopefully temporary. I don't like using them since they obscure control flow as they get called at some later time potentially by
+// a different thread. The threading issue doesn't matter at the moment since everything important runs on one thread but this could change if we want
+// to have separate UI and engine threads (as I believe we should). A better design would be to use async & await.
+
+namespace C7Engine
+{
+using System.Collections.Generic;
+using System.Linq;
+using C7GameData;
+
+public class AnimationTracker {
+	public delegate bool OnAnimationCompleted(string unitGUID, MapUnit.AnimatedAction action);
+
+	public static readonly OnAnimationCompleted doNothing = (unitGUID, action) => { return true; };
+
+	public struct ActiveAnimation {
+		public ulong startTimeMS, endTimeMS;
+		public MapUnit.AnimatedAction action;
+		public OnAnimationCompleted callback;
+	}
+
+	private Dictionary<string, ActiveAnimation> activeAnims    = new Dictionary<string, ActiveAnimation>();
+	private Dictionary<string, ActiveAnimation> completedAnims = new Dictionary<string, ActiveAnimation>();
+
+	public void startAnimation(ulong currentTimeMS, string unitGUID, MapUnit.AnimatedAction action, OnAnimationCompleted callback)
+	{
+		ulong animDurationMS = 500; // Hard-code durations to 0.5 sec for now. Ultimately we'll want to figure this out based on the INI file.
+
+		ActiveAnimation aa;
+		if (activeAnims.TryGetValue(unitGUID, out aa)) {
+			// If there's already an animation playing for this unit, end it first before replacing it
+			aa.callback(unitGUID, aa.action);
+		}
+		aa = new ActiveAnimation { startTimeMS = currentTimeMS, endTimeMS = currentTimeMS + animDurationMS, action = action, callback = callback ?? doNothing };
+
+		activeAnims[unitGUID] = aa;
+		completedAnims.Remove(unitGUID);
+	}
+
+	public void endAnimation(string unitGUID, bool triggerCallback = true)
+	{
+		ActiveAnimation aa;
+		if (triggerCallback && activeAnims.TryGetValue(unitGUID, out aa)) {
+			var forget = aa.callback(unitGUID, aa.action);
+			if (! forget)
+				completedAnims[unitGUID] = aa;
+			activeAnims.Remove(unitGUID);
+		} else {
+			activeAnims   .Remove(unitGUID);
+			completedAnims.Remove(unitGUID);
+		}
+	}
+
+	public bool hasCurrentAction(string unitGUID)
+	{
+		return activeAnims.ContainsKey(unitGUID) || completedAnims.ContainsKey(unitGUID);
+	}
+
+	public (MapUnit.AnimatedAction, double) getCurrentActionAndPeriod(string unitGUID, ulong currentTimeMS)
+	{
+		ActiveAnimation aa;
+		if (! activeAnims.TryGetValue(unitGUID, out aa))
+			aa = completedAnims[unitGUID];
+
+		var durationMS = (double)(aa.endTimeMS - aa.startTimeMS);
+		if (durationMS <= 0.0)
+			durationMS = 1.0;
+		var period = (double)(currentTimeMS - aa.startTimeMS) / durationMS;
```

Comment:
If I'm reading the math right, isn't the `period` variable actually the percentage of the way through the animation that the code would be?  E.g. if startTimeMS is 100, and endTimeMS is 400, and currentTimeMS is 200, durationMS would be 300, and period would be 100/300 = 1/3.

If so, I don't think `period` is the right name for it, as that suggests a recurring interval.  The `progress` variable used at a higher level than this in UnitInteractions.cs seems like a more accurate name.

Reply:
I was using "progress" to store only values 0.0 to 1.0 so I wanted a different name for that variable. "period" isn't ideal, but it sort of makes sense if you think of it as the number of periods the animation has gone through, but it's a floating point so it stores fractional periods as well. I'll rename it to "repetitionCount" for clarity.